internal/sync/lifecycle.go

```diff
@@ -117,39 +101,45 @@ func Once() LifecycleOnce {
 // If Start is called multiple times it will return the error
 // from the first time it was called.
 func (l *lifecycleOnce) Start(f func() error) error {
-	if l.starting.Swap(true) {
-		<-l.startCh
+	if l.state.CAS(int32(Idle), int32(Starting)) {
+		if f != nil {
+			l.err = f()
+		}
+
+		// skip forward to error state
+		if l.err != nil {
+			l.state.Store(int32(Errored))
```

Comment:
For the record, I convinced myself it was okay to Store and not CAS because only one thread can win the race to run the idle->starting state transition, and all others, including Stop(), block on startCh.